### DISCUSSION TITLE
Potentially introduce a new PixelColorBlendingMode.And value

### DISCUSSION THREAD

**Discussion by JimmyWoodhouse** [SixLabors/ImageSharp.Drawing#341] [Ideas]
It seems that the functionality for does not currently exist for drawing onto an image and only keeping the overlapping sections of the image with a polygon, or similar.

For example, my use case is that I have an image where I wish to make it transparent in a certain area. However, the contents of this image have a stroke/outline, so the stroke needs to be repositioned.
My idea was to clone the image and draw a polygon onto the cloned image, only take the overlapping areas. I would then draw the overlapping areas (from the cloned image) onto the original image in a different colour.

In regards to a possible new PixelColorBlendingMode.And with my examples below, the image would *only* contain either a red or a blue right angled triangle where they overlap (potentially even choosing between new PixelColorBlendingMode.SrcAnd and PixelColorBlendingMode.DestAnd values for this)

Sorry if I'm wrong here, but I'm not sure if the current PixelColorBlendingMode values all work correctly? I might be wrong and misunderstanding them here though, but here is my testing:

Image:
![1](https://github.com/user-attachments/assets/11ac36ee-8bec-4447-9909-47f0153ae907)

Polygon that gets drawn onto the image:
![2](https://github.com/user-attachments/assets/6afab089-c314-4103-814f-325f02fbcfd9)

Here are the results:
![image](https://github.com/user-attachments/assets/5bb613c8-a359-4117-a3ad-d4dc4570928d)

I thought you'd have received something very different for either DestAtop or DestIn:
![ffff](https://github.com/user-attachments/assets/5b5da9a9-3a9f-4019-906f-9373842ee3f2)

Thanks!

**Comment by JimmyWoodhouse** [comment]
To add to this, here's some examples:

```
var drawingOptions = new DrawingOptions
{
    GraphicsOptions = new GraphicsOptions()
    {
        AlphaCompositionMode = PixelAlphaCompositionMode.SrcAnd
    }
};

var brush = new SolidBrush(Color.Red);
var points = new PointF[] { new PointF(0, 0), new PointF(512, 0), new PointF(512, 512), new PointF(0, 512) };

image.Mutate(i => i.FillPolygon(drawingOptions, brush, polygon.Points.ToArray()));
```

using the red and blue images above, you would get:
![SrcAnd](https://github.com/user-attachments/assets/8e127f7e-f9d1-497a-b288-628e1e621ab6)

and for DestAnd:
![DestAnd](https://github.com/user-attachments/assets/18cc0b9d-33dc-4ad0-a0a2-4f2ea5fc4549)

**Comment by JimBobSquarePants** [comment]
@JimmyWoodhouse 

There's something not right about your results there as they do not match the expected output. 

Here's example code of mine which generates each entry.

```c#
using Image<Rgba32> src = Image.Load<Rgba32>(@"C:\Users\james\Downloads\src.png");
using Image<Rgba32> dst = Image.Load<Rgba32>(@"C:\Users\james\Downloads\dst.png");

foreach (var mode in Enum.GetValues<PixelAlphaCompositionMode>())
{
    GraphicsOptions options = new()
    {
        AlphaCompositionMode = mode
    };
    
    using Image<Rgba32> res = dst.Clone(x => x.DrawImage(src, options));
    res.Save(@$"C:\Users\james\Downloads\341-porter-duff\{mode}.png");
}

```

And here's the output. (Ignore Clear, Windows 11 has a bug where explorer thumbnails of fully clear PNG files sometimes display as black)

![image](https://github.com/user-attachments/assets/889143a0-4c3c-4e03-a8a5-e460d969626f)

Which matches the examples from this article on Porter-Duff Compositing Operators from the University of Auckland.

https://www.stat.auckland.ac.nz/~paul/Reports/GraphicsEngine/compositing/compositing.html

`DestIn` and `SrcIn` already exist providing the results that you require.

**Reply by JimmyWoodhouse** [reply]
Your code given works for me too, so it's correct when drawing images onto other images.
However, it looks like it is broken when filling polygons onto images.

Here are my results when using the same Src image above, and the points that make up the Dest image above (but not the Dest image itself)

![image](https://github.com/user-attachments/assets/7a2734df-dd70-4fd4-a36c-1ee8606de6eb)

Here's the code I used for this:

```
using Image<Rgba32> src = Image.Load<Rgba32>(@"C:\Users\Jimmy\Downloads\testingimages\src.png");

var drawingOptions = new DrawingOptions
{
    GraphicsOptions = new GraphicsOptions()
    {
        AlphaCompositionMode = PixelAlphaCompositionMode.SrcIn
        //AlphaCompositionMode = PixelAlphaCompositionMode.DestIn
    }
};

var brush = new SolidBrush(Color.CornflowerBlue);
var points = new PointF[] { new PointF(0, 0), new PointF(512, 0), new PointF(512, 512), new PointF(0, 512) };

src.Mutate(i => i.FillPolygon(drawingOptions, brush, points.ToArray()));


src.SaveAsPng(@"C:\Users\Jimmy\Downloads\testingimages\srcIn.png");
//src.SaveAsPng(@"C:\Users\Jimmy\Downloads\testingimages\destIn.png");
```

Good to know that DestIn and SrcIn provide this though, I thought it was strange that it didn't seem to have anything for those scenarios!

**Reply by JimBobSquarePants** [reply]
I'm seeing the following:

![image](https://github.com/user-attachments/assets/a042ab6f-ad10-4ed9-aee9-1c4e3f7d5d1a)

The algorithm is drawing the parts of the polygon as specified by the composition mode. 
I think it's correct, but I've not had enough sleep to be 100% confident. It's probably a good idea to compare them to a skiasharp equivalent.

```c#
using Image<Rgba32> dst = Image.Load<Rgba32>(@"C:\Users\james\Downloads\dst.png");
SolidBrush brush = new(Color.CornflowerBlue);
PointF[] points = [new PointF(0, 0), new PointF(512, 0), new PointF(512, 512), new PointF(0, 512)];

foreach (var mode in Enum.GetValues<PixelAlphaCompositionMode>())
{
    DrawingOptions options = new()
    {
        GraphicsOptions = new()
        {
            AlphaCompositionMode = mode
        }
    };

    using Image<Rgba32> res = dst.Clone(x => x.FillPolygon(options, brush, points));
    res.Save(@$"C:\Users\james\Downloads\341-porter-duff\poly-{mode}.png");
}
```

**Reply by JimmyWoodhouse** [reply]
You're getting the same results as me here it seems, but it seems wrong to me haha.
SrcIn acts the same as SrcAtop, DestIn acts the same as Dest, and Src isn't the source image only?
There is no "and" combination in your results when drawing a PointF[] here.
Might be me misunderstanding obviously but it doesn't seem right to me!

**Reply by JimBobSquarePants** [reply]
LGTM.

![image](https://github.com/user-attachments/assets/31564eeb-aec3-46ee-b33e-ca4460f12324)

```c#
string imagePath = @"C:\Users\james\Downloads\dst.png";
string outputPath = @"C:\Users\james\Downloads\341-porter-duff\skia\";

// Load the source image into SkiaSharp
using SKBitmap dstBitmap = SKBitmap.Decode(imagePath);
SKImageInfo imageInfo = new SKImageInfo(dstBitmap.Width, dstBitmap.Height);

// Define the polygon points
SKPoint[] points = new SKPoint[]
{
    new SKPoint(0, 0),
    new SKPoint(512, 0),
    new SKPoint(512, 512),
    new SKPoint(0, 512)
};

// Define the brush color
SKColor brushColor = SKColors.CornflowerBlue;

SKBlendMode[] porterDuffModes = new[]
{
    SKBlendMode.Clear,
    SKBlendMode.Src,
    SKBlendMode.Dst,
    SKBlendMode.SrcOver,
    SKBlendMode.DstOver,
    SKBlendMode.SrcIn,
    SKBlendMode.DstIn,
    SKBlendMode.SrcOut,
    SKBlendMode.DstOut,
    SKBlendMode.SrcATop,
    SKBlendMode.DstATop,
    SKBlendMode.Xor
};

// Iterate through all SkBlendMode options
foreach (SKBlendMode mode in porterDuffModes)
{
    using SKSurface surface = SKSurface.Create(imageInfo);
    using SKCanvas canvas = surface.Canvas;

    // Draw the background image onto the canvas
    canvas.DrawBitmap(dstBitmap, 0, 0);

    // Set up the paint with the specified blend mode
    using SKPaint paint = new SKPaint
    {
        Color = brushColor,
        BlendMode = mode,
        IsAntialias = true
    };

    // Draw the polygon with the specified blend mode
    using SKPath path = new SKPath();
    path.AddPoly(points, close: true);
    canvas.DrawPath(path, paint);

    // Save the resulting image to a file
    using SKImage resultImage = surface.Snapshot();
    using SKData data = resultImage.Encode(SKEncodedImageFormat.Png, 100);
    string filePath = Path.Combine(outputPath, $"sk-poly-{mode}.png");
    Directory.CreateDirectory(outputPath);
    File.WriteAllBytes(filePath, data.ToArray());
}

```

**Reply by JimmyWoodhouse** [reply]
Cheers, I must have misinterpreted what was going on.
I was expecting your results with using images, to match with your results when using polygons:

![image](https://github.com/user-attachments/assets/28a0a09f-3ddc-402c-ab5e-65cf43e80f6e)

![image](https://github.com/user-attachments/assets/ffe4b613-9bb0-4a84-818e-c038da1e54a1)

This (the result when using images) is what I was hoping to get when using polygons:
![image](https://github.com/user-attachments/assets/dfd0f9c8-499d-4f9b-b46d-c5ddb8328227)

But you and I get this when using polygons:
![image](https://github.com/user-attachments/assets/e45d1ef8-0610-4f11-8d91-596fba945468)

So yeah, I assume that's a misunderstanding on how the process works on my end.

Thanks for the explanation
